### PR TITLE
[ISSUE-1261] set_finished LocalExchangeSourceOperator because of short-circuit break of HashJoinProbeOperator

### LINKED ISSUE #1261: set_finished LocalExchangeSourceOperator because of short-circuit break of HashJoinProbeOperator
## Bugfix

In a pipeline, some operators can finish in advance, e.g.
1. LimitOperator: it finishes after consume sufficient rows.
2. HashJoinProbeOperator: it finishes if HashJoinBuildOperator build a empty HT and the join type is INNER JOIN, RIGHT JOIN or RIGHT SEMI-JOIN.

when a pipeline is in the form of  ExchangeSourceOperator->Operator(finish in advance) and the pipeline's dop > 1, LocalExchange would get stuck in a corner case if:
1. There exists a driver, the latter operator(i.e. HashJoinProbeOperator) is finished, then finish function of the former ExchangeSourceOperator ignores the buffered chunks the number of rows of that is taken into consideration when LocalExchanger determines whether it has available room to accept new chunks from LocalExchangeSinkOperators.
2. LocalExchanger have no available room to accept new chunks from LocalExchangeSinkOperators, so drivers that ends with the LocalExchangeSinkOperator are blocked in OUTPUT_FULL state, and drivers that starts with LocalExchangeSourceOperator are blocked in INTPUT_EMPTY state, in such scenarios, a query is always executed timeout.

The solution to this issue is that set_finished function which is invoked in the timepoint distinct from the set_finishing(original name is Operator::finish) function is added to Operator.  set_finished function is used to shutdown both input/output stream and clear the buffered data inside, while set_finishing function is just used to shutdown input stream of the operator that becomes finished until it has processed all buffered data inside.

Most of operators have trivial set_finished function that just call set_finishing function. however, ExchangeSourceOperator's set_finished function will clear the buffered chunks and subtract the number of rows of the buffered chunks from LocalExchanger and make it unblocked.

### PR DESCRIPTION
**Pull request by satanson**
## Bugfix
In a pipeline, some operators can finish in advance, e.g.
1. LimitOperator: it finishes after consume sufficient rows.
2. HashJoinProbeOperator: it finishes if HashJoinBuildOperator build a empty HT and the join type is INNER JOIN, RIGHT JOIN or RIGHT SEMI-JOIN.

when a pipeline is in the form of  ExchangeSourceOperator->Operator(finish in advance) and the pipeline's dop > 1, LocalExchange would get stuck in a corner case if:
1. There exists a driver, the latter operator(i.e. HashJoinProbeOperator) is finished, then finish function of the former ExchangeSourceOperator ignores the buffered chunks the number of rows of that is taken into consideration when LocalExchanger determines whether it has available room to accept new chunks from LocalExchangeSinkOperators.
2. LocalExchanger have no available room to accept new chunks from LocalExchangeSinkOperators, so drivers that ends with the LocalExchangeSinkOperator are blocked in OUTPUT_FULL state, and drivers that starts with LocalExchangeSourceOperator are blocked in INTPUT_EMPTY state, in such scenarios, a query is always executed timeout.

The solution to this issue is that set_finished function which is invoked in the timepoint distinct from the set_finishing(original name is Operator::finish) function is added to Operator.  set_finished function is used to shutdown both input/output stream and clear the buffered data inside, while set_finishing function is just used to shutdown input stream of the operator that becomes finished until it has processed all buffered data inside.

Most of operators have trivial set_finished function that just call set_finishing function. however, ExchangeSourceOperator's set_finished function will clear the buffered chunks and subtract the number of rows of the buffered chunks from LocalExchanger and make it unblocked.

### CODE REVIEW

**Comment by satanson** [comment]
run starrocks_be_unittest

File: be/src/exec/pipeline/operator.h [hunk 1]
```diff
@@ -43,8 +43,14 @@ class Operator {
...
     // The method should be idempotent, because it may be triggered
     // multiple times in the entire life cycle
+    // finish function is used to finish the following operator of the current operator that encounters its EOS
+    // and has no data to push into its following operator.
     virtual void finish(RuntimeState* state) = 0;
 
+    // finish_backward is used to finish the preceding operator of the current operator that finishes in advance
+    // and never need input data to consume.
+    virtual void finish_backward(RuntimeState* state) { finish(state); }
+
```
Comment (liuyehcf) [inline be/src/exec/pipeline/operator.h:53]: Why not merge those two methods into one? Derived class can have different meanings.

Comment (satanson) [reply be/src/exec/pipeline/operator.h:53]: two functions have different meanings:
1. finish func is invoked by the current operator to its following operator that the EOS is comming, but the following operator  should still process the buffered chunks inside.
2. finish_backward is invoked by the current operator to preceding operators that they should be finished, the buffered data is needless.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 2]
```diff
@@ -43,6 +49,28 @@ bool LocalExchangeSourceOperator::has_output() const {
            (_is_finished && _partition_rows_num > 0);
 }
 
+void LocalExchangeSourceOperator::finish_backward(RuntimeState* state) {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:52]: I think we no need to add a `finish_backward` method, you could implement this logic in `finish`.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 3]
```diff
@@ -43,6 +49,28 @@ bool LocalExchangeSourceOperator::has_output() const {
...
+    _is_finished = true;
+    // Compute out the number of rows of the _full_chunk_queue.
+    size_t full_rows_num = 0;
+    while (!_full_chunk_queue.empty()) {
+        auto&& chunk = std::move(_full_chunk_queue.front());
+        _full_chunk_queue.pop();
+        full_rows_num += chunk->num_rows();
+    }
+    // clear _full_chunk_queue
+    { [[maybe_unused]] typeof(_full_chunk_queue) tmp = std::move(_full_chunk_queue); }
```
Comment (kangkaisen) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:66]: Why use this strange way to clear the queue?

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp (hunk 2 above)
Comment (satanson) [reply be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:52]: finish func is invoked by the current operator to its following operator that the EOS is comming, but the following operator should still process the buffered chunks inside.
finish_backward is invoked by the current operator to preceding operators that they should be finished, the buffered data is needless.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp (hunk 3 above)
Comment (satanson) [reply be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:66]: queue have no clear function

**Review by kangkaisen — APPROVED** [review]

**Review by liuyehcf — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Review by Pslydhh — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**